File: nlp4bpm_commons/src/main/java/edu/upc/nlp4bpm_commons/FreelingAPI.java
package edu.upc.nlp4bpm_commons;

import java.util.Arrays;
import java.util.List;

import clojure.java.api.Clojure;
import clojure.lang.Keyword;
import clojure.lang.IFn;

public class FreelingAPI {

    static {
        IFn require = Clojure.var("clojure.core", "require");
        require.invoke(Clojure.read("edu.upc.nlp4bpm-commons.freeling-api"));
        require.invoke(Clojure.read("edu.upc.nlp4bpm-commons.config"));
    }

    public static String analyze (String text, String lang, String output, String level) {
        IFn analyze = Clojure.var("edu.upc.nlp4bpm-commons.freeling-api", "analyze");
        return (String) analyze.invoke(Keyword.intern("text"), text,
                                       Keyword.intern("lang"), lang,
                                       Keyword.intern("output"), output,
                                       Keyword.intern("level"), level);
    }

    public static String analyzeCached (String text, String lang, String output, String level) {
        IFn analyzeCached = Clojure.var("edu.upc.nlp4bpm-commons.freeling-api", "analyze-cached");
        return (String) analyzeCached.invoke(Keyword.intern("text"), text,
                                             Keyword.intern("lang"), lang,
                                             Keyword.intern("output"), output,
                                             Keyword.intern("level"), level);
    }

    public static String analyzeCachedIfAvailable(String text, String lang, String output, String level) {
        IFn analyzeCached = Clojure.var("edu.upc.nlp4bpm-commons.freeling-api", "analyze-cached-if-available");
        return (String) analyzeCached.invoke(Keyword.intern("text"), text,
                                             Keyword.intern("lang"), lang,
                                             Keyword.intern("output"), output,
                                             Keyword.intern("level"), level);
    }

    /**
       Analyzes a list of labels as string. Returns a list of analyzed labels (see AnalyzedLabel.java for structure details)
       The analysis can be one of: "task", "swimlane" or "other". The analysis type of task is for task labels, swimlane is for
       swimlane/pool labels while type other performs only a token-based analysis without parsing or predicate extraction.
       TODO: WIP: Add "swimlane" analysis type
       TODO: Only works for english
     **/
    @SuppressWarnings("unchecked")
	public static List<AnalyzedLabel> analyzeLabels (List<String> labels, String language, String analysis_type) {
        IFn analyzeLabels = Clojure.var("edu.upc.nlp4bpm-commons.freeling-api", "analyze-labels-java");
        return (List<AnalyzedLabel>) analyzeLabels.invoke(Keyword.intern("labels"), labels,
                                             Keyword.intern("lang"), language,
                                             Keyword.intern("label-type"), analysis_type);
    }

    //NOTE: analyzedLabelsCached cannot be implemented b.c. freeling "Sentence"s cannot be serialized.

    public static void setCredentialsPath (String path) {
        IFn setCredentialsPath = Clojure.var("edu.upc.nlp4bpm-commons.freeling-api", "set-credentials-path!");
        setCredentialsPath.invoke(path);
    }

    public static void setConfig (String key, String val) {
        IFn setConfig = Clojure.var("edu.upc.nlp4bpm-commons.config", "-setConfig");
        setConfig.invoke(key, val);
    }

    public static void getConfig (String key) {
        IFn getConfig = Clojure.var("edu.upc.nlp4bpm-commons.config", "-getConfig");
        getConfig.invoke(key);
    }

    public static void setMode (String mode) {
        IFn setMode = Clojure.var("edu.upc.nlp4bpm-commons.freeling-api", "set-mode");
        setMode.invoke(mode);
    }
    
    public static void main (String[] args) {
    	// Sample input data
    	List<String> labels = Arrays.asList("Deliver customer card", "Process blue form", "Instruct employees");
    	
    	// Call the analyzer module
    	List<AnalyzedLabel> l = analyzeLabels(labels, "en", "task");
    	
    	// Get the results
    	int tk_idx = l.get(0).getHeadOf("predicate"); //Tokens are returned as indices
    	List<Integer> word_idxs = l.get(0).getWordsOf("predicate");
    	edu.upc.Jfreeling.Word w = l.get(0).getSentence().getWords().get(tk_idx); //You can retrieve the token data from the sentence
    	System.out.println(w.getLemma());
    }
}

File: nlp4bpm_commons/src/main/clojure/edu/upc/nlp4bpm_commons/freeling_api.clj
(ns edu.upc.nlp4bpm-commons.freeling-api
  (:require [edu.upc.nlp4bpm-commons.textserver :as textserver]
            [edu.upc.nlp4bpm-commons.config :as config]
            [edu.upc.nlp4bpm-commons.Freeling :as freeling]
            [edu.upc.nlp4bpm-commons.label-parser :as label-parser]
            [edu.upc.nlp4bpm-commons.cache :as cache]))

(defn set-mode [the-mode]
  (assert (#{"online" "local"} the-mode) "Mode must be one of: online, local")
  (config/-setConfig "mode" the-mode))

(defn analyze [& {:keys [text lang output level mode-override]
                  :or   {output "json", lang "en", level "semgraph"}}]
  (let [mode (or mode-override (config/-getConfig "mode"))]
    (cond
      (= mode "online") (textserver/analyze :text text :output output :lang lang :level level)
      (= mode "local")  (freeling/analyze :text text :output output :lang lang :level level))))

(defn analyze-labels-java [& {:keys [labels lang label-type]}]
  (label-parser/parse-labels-java
   labels lang
   :label-type label-type))

(defn analyze-labels [& {:keys [labels lang label-type start-id]
                         :or {lang "en" labels []
                              label-type "task" start-id 1}}]
  (label-parser/parse-labels
   labels lang
   :label-type label-type :start-id start-id))

(defn analyze-labels-cached [& {:keys [labels lang label-type start-id]
                                :or {lang "en" labels []
                                     label-type "task" start-id 1}}]
  (if-let [maybe-response (cache/retrieveItem [labels lang label-type start-id])]
    maybe-response
    (let [response (analyze-labels :labels labels :lang lang :label-type
                                   label-type :start-id start-id)]
      (cache/addItemIfNotExists [labels lang label-type start-id] response)
      response)))

(defn analyze-cached [& {:keys [text lang output level]
                         :or {output "json", lang "en", level "semgraph"}}]
  "Same as analyze, cached version."
  (if-let [maybe-response (cache/retrieveItem [text lang output level])]
    maybe-response
    (let [response (analyze :text text :lang lang :output output :level level)]
      (cache/addItemIfNotExists [text lang output level] response)
      response)))

(defn analyze-cached-if-available [& {:keys [text lang output level]
                                      :or {output "json", lang "en", level "semgraph"}}]
  (cache/if-cache
   (analyze-cached :text text :lang lang :output output :level level)
   (analyze :text text :lang lang :output output :level level)))

(defn set-credentials-path! [path]
  (swap! edu.upc.nlp4bpm-commons.textserver/credentials-path (constantly path)))

